docs/manual/resource-leak-checker.tex

```diff
@@ -588,6 +597,234 @@
 will only use \<@MustCallUnknown> as a default when encountering a language
 feature that it is unable to reason about.
 
+\sectionAndLabel{Arrays of resources}{resource-leak-arrays}
+To understand how the checker handles arrays, let us first remind ourselves what happens if we assign resources with non-empty must-call type to an array:
+
+\begin{verbatim}
+Socket[] s = new Socket[n];
+s[0] = new Socket(myHost, myPort);
+\end{verbatim}
+
+The assignment throws an error, because the scope of the constructed Socket ends, without its must-call obligations having been fulfilled.
+
+\begin{verbatim}
+Example.java:31: error: [required.method.not.called] @MustCall method close may not have been invoked on new Socket(myHost, myPort) or any of its aliases.
+            arr[0] = new Socket(myHost, myPort);
+                     ^
+  The type of object is: java.net.Socket.
+  Reason for going out of scope: regular method exit
+1 error
+\end{verbatim}
+
+We need a way for the array to take on this obligation for the resource. This functionality is jointly provided by the \<MustCallOnElements> and \<CalledMethodsOnElements> subcheckers.
+
+\subsectionAndLabel{Declaration and assignment of resource-holding arrays}{resource-leak-arrays-declaration}
+Annotating the array with the \<@OwningArray> declaration annotation allows it to take on such obligations by using a pattern-matched assignment syntax:
+
+\begin{verbatim}
+@OwningArray Socket[] s = new Socket[n];
+// 1. @MustCallOnElements({}) @CalledMethodsOnElements({})
+for (int i = 0; i < s.length; i++) {
+  try {
+    s[i] = new Socket(myHost, myPort);
+  } catch (Exception e) {
+    // do nothing
+  }
+}
+// 2. @MustCallOnElements({"close"}) @CalledMethodsOnElements({})
+\end{verbatim}
+The loop must iterate from 0 to \<s.length> or \<n>, where \<n> is the effectively final variable used for the initialization of the array size. The update must be either a pre-increment or post-increment.  There must be exactly one statement in the body (which may be a try-catch construct) that contains the assignment. The right-hand side of the assignment must be the immediate construction of the object (not simply any expression evaluating to one). If there is a try-catch construct, the catch and finally blocks may not throw any exceptions, break out of the loop, call any methods or return for successful pattern-matching. In order to assign the elements of an \<@OwningArray>, the array must have no open calling obligations on its elements.
```

Comment:
When you say "no open calling obligations" in the last sentence, does that mean the same thing as saying the array variable must be `@MustCallOnElements({})` just before the loop?

Reply:
it means the mcoe values must be a subset of the cmoe values just before the loop. I added it.